src/Processors/Formats/Impl/JSONColumnsWithMetadataBlockOutputFormat.cpp

```diff
@@ -0,0 +1,120 @@
+#include <Processors/Formats/Impl/JSONColumnsWithMetadataBlockOutputFormat.h>
+#include <Formats/JSONUtils.h>
+#include <Formats/FormatFactory.h>
+#include <IO/WriteHelpers.h>
+#include <IO/WriteBufferValidUTF8.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+JSONColumnsWithMetadataBlockOutputFormat::JSONColumnsWithMetadataBlockOutputFormat(WriteBuffer & out_, const Block & header_, const FormatSettings & format_settings_)
+    : JSONColumnsBlockOutputFormat(out_, header_, format_settings_, 1)
+{
+    bool need_validate_utf8 = false;
+    JSONUtils::makeNamesAndTypesWithValidUTF8(fields, format_settings, need_validate_utf8);
+
+    if (need_validate_utf8)
+    {
+        validating_ostr = std::make_unique<WriteBufferValidUTF8>(out);
+        ostr = validating_ostr.get();
+    }
+}
+
+void JSONColumnsWithMetadataBlockOutputFormat::writePrefix()
+{
+    JSONUtils::writeObjectStart(*ostr);
```

Comment:
Just to check and I am not 100% sure but for most of the methods in JSONColumnsWithMetadataBlockOutputFormat, I would have expected calls to the base class JSONColumnsBlockOutputFormat or no implementation at all (so that the base class method is called instead).

Writing the actual data is handled by the base class and the subclass in theory only adds the statistics/metadata?

Reply:
JSONColumnsBaseBlockOutputFormat class is responsible for writing data in the next format:
```
<chunk start specified by subclass>
<column start specified by subclass><column data><column end specified by subclass>
...
<chunk end specified by subclass>
```
In JSONColumns and JSONColumnsWithMetadata we have the same format of writing columns, so in JSONColumnsWithMetadataBlockOutputFormat parent class JSONColumnsBlockOutputFormat is responsible for specifying all necessary stuff for writing columns, but in JSONColumnsWithMetadataBlockOutputFormat we add some extra data like metadata/statistics and we should do it before and after this chunk, so we implement special methods from base class IOutputFormat - `writePreffix` for writing metadata before chunk and `finalizeImpl` for writing statistics.